heralding/capabilities/smtp.py

```diff
@@ -21,89 +21,251 @@
 # display, publicly perform, sublicense, relicense, and distribute [the] Contributions
 # and such derivative works.
 
-
-
+import time
+import random
+import base64
+import socket
+import asyncio
 import logging
-from base64 import b64decode
+import aiosmtpd
 
 from aiosmtpd.smtp import SMTP, MISSING, syntax
 
 from heralding.capabilities.handlerbase import HandlerBase
 
-EMPTYSTRING = ""
-logger = logging.getLogger(__name__)
+log = logging.getLogger(__name__)
+aiosmtpd.smtp.log = log
 
 
 class SMTPHandler(SMTP):
-    def __init__(self, reader, writer, options, session):
-        super().__init__(None)
-        self.transport = writer
-        self._writer = writer
+    def __init__(self, reader, writer, session, options, loop):
+        self.banner = options['protocol_specific_data']['banner']
+        super().__init__(None, hostname=self.banner, loop=loop)
+        # Reset standard banner.
+        self.__ident__ = ""
         self._reader = reader
+        self._writer = writer
+        self.transport = writer
+        self.fqdn = socket.getfqdn()
+
+        self._set_rset_state()
         self.session = session
+        self.session.peer = self.transport.get_extra_info('peername')
         self.session.extended_smtp = None
         self.session.host_name = None
-        self.max_tries = int(options['protocol_specific_data']['max_attempts'])
-        self.banner = options['protocol_specific_data']['banner']
 
-    @syntax('HELO hostname')
-    async def smtp_EHLO(self, hostname):
-        await super().smtp_EHLO(hostname)
-        await self.push('250-AUTH PLAIN')
+    async def push(self, status):
+        response = bytes(
+            status + '\r\n', 'utf-8' if self.enable_SMTPUTF8 else 'ascii')
+        self._writer.write(response)
```

Comment:
Blocking is deffered until `drain()`. But as said in docs (https://docs.python.org/3/library/asyncio-stream.html#asyncio.StreamWriter.drain), if there is nothing to wait for, the yield-from (await) continues immediately.